posts/inside-rust/2022-04-20-libs-aspirations.md

```diff
@@ -0,0 +1,209 @@
+Unlike most crates, we cannot release a new major version, as that would effectively be releasing a 'Rust 2.0'.
+So, once an API is stable, we have to keep it there forever, meaning that we have to be extremely careful
+when stabilizing anything new.
+
+While we are very careful with adding new APIs, mistakes can still happen.
+There are a few things we would do differently if we could back in time, based on current experience.
+There are not a lot of these cases, but over time these can still accumulate to the point that
+it'd be useful to have a mechanism to correct past mistakes.
+
+The Rust language has the concept of [editions](https://doc.rust-lang.org/edition-guide/editions/index.html)
+to be able to make breaking changes in the language itself, without disrupting Rust users.
+The library, however, can make very limited use of editions to correct mistakes.
+We have used them for the [`panic!()` macro](https://doc.rust-lang.org/edition-guide/rust-2021/panic-macro-consistency.html)
+and [the prelude](https://doc.rust-lang.org/edition-guide/rust-2021/prelude.html).
+However, in general, it's extremely tricky to make use of the edition mechanism for backwards incompatible
+library changes, as crates of different editions can be mixed, while all using the same standard library.
+
+There are also cases when adding a new API can subtly break existing Rust code,
+even when not changing any existing API.
+This can happen, for example, when a type gets a new method that was already available through
+a popular trait, such as `itertools::Itertools` or `num::Integer`.
+Adding a method to the standard library can result in an existing method call resolving differently,
+potentially breaking code.
+This is usually considered 'acceptable' breakage, but as Rust usage grows,
+the impact of such breakage increases, often making such breakage unacceptable in practice.
+
+So, in order to keep evolving the standard library, we'd like to collaborate on language features
+that alleviate these issues:
+
+- [Edition based method disambiguation](https://github.com/rust-lang/rfcs/pull/3240)
+- A way to fix the `Range` types, such that `1..2` can be `Copy`.
+- A way to remove or improve lock poisoning without breaking existing code.
+- General mechanisms to provide for the library what editions provide for the language.
+
+**People and collaboration**
+
+The most important thing to keep Rust and the ecosystem scalable,
+are the people: Rust team members, maintainers of crates in the ecosystem,
+reviewers, contributors, and so on.
+It's important we keep working on how we collaborate and make it
+as easy as possible for everyone to get involved in a way that works for them.
+
+Concretely, we want to work on:
+
+- Better and more complete guidelines for contributors and reviewers; and
+- More interaction with the rest of the ecosystem.
+
+**Making `std` less special / Empowering other crates in the ecosystem**
```

Comment:
I'm not convinced this one is a good idea. It's good that at least std gets to "dogfood" these features, as it were. It also is safer (err, not in a safety sense, but you know) for std to rely on them than anybody else, because it knows the precise version of the compiler it ships with, which isn't the case for crates on crates.io.

Perhaps it should be reduced some, though? And we should probably push towards making more of the features we use in std stable.

Reply:
the compiler is also a good place to dogfood these features as well fwiw

Reply:
(We discussed this live the meeting just now. Summary for those reading along: rustc is a better place for dogfooding, and we care about the standard library being simple readable code, to serve as an example when people click `[source]` in rustdoc.)

Reply:
(And also: it's unclear how feasible the end goal is. But this is a direction we want to go towards and see what we bump into on the way, not a decision about the final state.)

Reply:
> (We discussed this live the meeting just now. Summary for those reading along: rustc is a better place for dogfooding, and we care about the standard library being simple readable code, to serve as an example when people click [source] in rustdoc.)

Strong agreement. Sounds amazing, frankly!